packages/core/src/utils/retry.ts

```diff
@@ -80,10 +81,31 @@ function defaultShouldRetry(
  * @param ms The number of milliseconds to delay.
  * @returns A promise that resolves after the delay.
  */
-function delay(ms: number): Promise<void> {
-  return new Promise((resolve) => setTimeout(resolve, ms));
-}
+function delay(ms: number, signal?: AbortSignal): Promise<void> {
+  return new Promise((resolve, reject) => {
+    const abortError = new Error('Aborted');
```

Comment:
nit: simplify the flow by having a trivial version if there is no signal and this complex logic if there is a signal.